2단계 - HTTP 웹 서버 구현

catsbi opened the pull request:
회사 스프린트와 병행하려니 매일 철야하면서 작업하는데도, 미션 진행이 잘 안되네요 ㅠㅠ.. 
테스트코드도 거의 없고, 뭔가 많이 부족하지만, 우선 리뷰를 받고 피드백 적용하면서 계속 진행하려고 우선 리뷰요청 드립니다!

Q1. 정적파일(js 등)을 응답할 때 별 특이사항이 없고 try-with-resources를 이용해서 자동 close를 하고 있는데 아래 첨부한 이미지처럼 응답결과에 header 내용이 들어가 있거나 다른 요청정보에 사용된 헤더파일이 작성되는 현상이 종종 있습니다.  요청마다 새로운 인스턴스를 생성하기에 정적 자원 공유는 아닌것 같은데 왜이럴까요? 
![image](https://user-images.githubusercontent.com/43158866/180820115-9d1ab87c-6a30-46a9-a5ad-e4ecc7e1d90e.png)


Q2. 페이지 요청시`Software caused connection abort: socket write error` 예외가 빈번하게 발생하는데, 어떻게 해결할지 모르겠네요..

catsbi committed:
feat: 프로토콜 정보를 래핑하는 일급 객체 생성 및 테스트 케이스 작성

catsbi committed:
refactor: import구문 수정 및 불변객체 설계 및 Json파싱 로직 수정

catsbi committed:
doc: 2단계 미션 기능 요구사항 문서화 정리

catsbi committed:
refactor: Http Method가 Post일 경우에만 Request Body 값 추출하여 생성하도록 로직 개선

catsbi committed:
feat: HttpHeaders 테스트 케이스 추가 및 로직 개선

catsbi committed:
feat: FrontController 테스트케이스 작성 및 기본 골격 생성

catsbi committed:
feat: FrontController 로직 개선

catsbi committed:
feat: FrontController 로직 개선

catsbi committed:
feat: FrontController에 메서드 실행 로직 추가

catsbi committed:
feat: step2 진행중

catsbi committed:
feat: step2 진행중

yongdae commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -9,53 +21,82 @@
+        responseBody(dos, httpResponse);
+    }
+
+    private void sendResponseHeader(DataOutputStream dos, HttpResponse httpResponse) {
```
Q1. 정적파일(js 등)을 응답할 때 별 특이사항이 없고 try-with-resources를 이용해서 자동 close를 하고 있는데 아래 첨부한 이미지처럼 응답결과에 header 내용이 들어가 있거나 다른 요청정보에 사용된 헤더파일이 작성되는 현상이 종종 있습니다. 요청마다 새로운 인스턴스를 생성하기에 정적 자원 공유는 아닌것 같은데 왜이럴까요?

<img width="1794" alt="스크린샷 2022-07-26 오후 10 21 17" src="https://user-images.githubusercontent.com/4317804/181016397-9c175f33-ff8e-42aa-a72c-a8bc0b9c7ea2.png">

A. toStringHeader()의 반환 값이 이상해요.
처음에는 StringBuilder를 초기화하는 `stringBuilder.setLength(0)` 코드를 의심했는데 이 부분은 문제가 되기 어렵다고 생각해서 조금 더 확인해보니 HttpResponse의 stringBuilder가 static으로 되어 있네요.

catsbi replied:
StringBuilder 의 setLength(0)은 StringBuilder를 정적으로 선언해서 재활용하기 위한 clear로직이였는데, 이게 문제가 맞았던 것 같습니다! StringBuilder 에서 setLength(0)으로 초기화를 시켜줘도 공유자원이라 ㅠㅠ.. 해결했습니다!

yongdae commented on src/main/java/webserver/WebApplicationServer.java:
```diff
@@ -1,17 +1,18 @@
 
-    public static void main(String args[]) throws Exception {
-        int port = 0;
+    public static void main(String[] args) throws Exception {
```
Q2. 페이지 요청시Software caused connection abort: socket write error 예외가 빈번하게 발생하는데, 어떻게 해결할지 모르겠네요..

A. 사용하고 있는 개발 환경에서는 동일한 이슈가 발생하지 않아 도움 드리기가 어려울 것 같아요.
조금 더 구체적인 상황을 설명해주시면 다시 확인해보도록 하겠습니다.

catsbi replied:
저도 StringBuilder 문제 하나 해결하고는 같이 해결되었습니다!

yongdae commented on src/main/java/webserver/config/WebConfig.java:
```diff
@@ -0,0 +1,34 @@
+import webserver.ui.UserController;
+import webserver.ui.WelcomeController;
+
+public class WebConfig {
```
WebConfig을 분리하신 이유가 궁금합니다. 🤔

catsbi replied:
구체화에 대한 의존성을 설정 클래스로 몰아서 관리하고자 한 것이였는데, 제대로역할을 못하고 있었네요. 
수정하겠습니다!

yongdae commented on src/main/java/webserver/ui/FrontController.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class FrontController {
```
FrontController는 Controller이라기보다는 Controller의 Container라고 생각되는데 클래스 이름을 변경해보면 어떨까요?

catsbi replied:
ControllerContainer라는 인터페이스로 핵심 업무를 모으고 ControllerContainerImpl이라는 구현체에 구현 로직을 작성했습니다!

yongdae commented on src/main/java/webserver/ui/Controller.java:
```diff
@@ -0,0 +1,37 @@
+    default Method getExecutableMethod(RequestLine requestLine) throws NoSuchMethodException {
+        return Arrays.stream(this.getClass().getDeclaredMethods())
+                .filter(method-> {
+                    RequestMapping annotation = method.getDeclaredAnnotation(RequestMapping.class);
+                    return Objects.nonNull(annotation)
+                            && supportMethod(method.getDeclaredAnnotation(RequestMapping.class), requestLine);
```
리플렉션을 활용해서 구현 해주셨네요. 💯 
이미 앞선 단계(@MVC 프레임워크 구현)의 미션을 수행해주셨네요. :)

yongdae commented on src/main/java/webserver/ui/Controller.java:
```diff
@@ -0,0 +1,37 @@
+package webserver.ui;
```
ui라는 패키지 이름을 사용하신 이유가 있으신가요?

catsbi replied:
컨트롤러는 UserInterface와 직접적으로 와닿는 계층이라고 생각해서 ui 패키지에 넣었습니다!

yongdae commented on src/main/java/webserver/domain/HttpResponse.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.function.Supplier;
+
+public class HttpResponse {
+    private static final StringBuilder stringBuilder = new StringBuilder();
+    private final HttpStatus httpStatus;
+    private final Attributes attributes = new Attributes();
+    private final HttpHeaders headers;
+
+    private final View view;
+
+    private final String body;
```
HttpResponse의 멤버 변수를 줄일 수 있는 방법을 고민해보면 좋을 것 같아요.

catsbi replied:
처음에는 Object body 타입하나를 가지고 핸들링을 하려고 했는데, 템플릿용과 아닌것에 대해서 고민하다보니 분리되서 관리되고 있어서 고민이네요.. 좀 더 고민해보겠습니다.

yongdae commented on src/main/java/webserver/domain/ContentType.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.stream.Collectors;
+
+public enum ContentType {
+    JS("application/javascript;charset=utf-8", "./static"),
+    HTML("text/html;charset=utf-8", "./templates"),
+    CSS("text/css;charset=utf-8", "./static"),
+    SVG("image/svg+xml;charset=utf-8", "./static"),
+    TTF("application/x-font-ttf;charset=utf-8", "./static"),
+    OTF("application/x-font-opentype;charset=utf-8", "./static"),
+    WOFF("application/font-woff;charset=utf-8", "./static"),
+    WOFF2("application/font-woff2;charset=utf-8", "./static"),
+    ICO( "image/avif;charset=utf-8", "./templates"),
+    JSON("application/json;charset=utf-8", "");
```
쉽지 않으셨을 텐데 ContentType을 일일히 정의해주셨네요. 👍

yongdae commented on src/main/java/webserver/domain/Cookie.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class Cookie {
```
아직 진행 중이라 없을 수 있는 내용인 것 같은데
Cookie에 대한 테스트 코드를 추가해보면 좋을 것 같아요.

catsbi replied:
테스트코드가 전체적으로 엄청 부족합니다. 빠르게 채워놓겠습니다!

yongdae commented on src/main/java/webserver/ui/UserController.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.Collections;
+import java.util.List;
+
+public class UserController implements Controller {
```
UserController를 사용한다면 요구사항을 기반으로 도출된 기능을 테스트 할 수 있지 않을까요? 🤔

yongdae replied:
UserController를 테스트에서 직접 호출해서 사용자 등록, 로그인, 사용자 목록 등의 기능을 검증 해보면 좋을 것 같아요.

catsbi replied:
api 테스트를 restTemplate같은 툴이 아니라 직접 호출하는 테스트도 작성을 하는게 기존 api 테스트와 비교해서 어떤 차이점이 있을까요?

yongdae replied:
restTemplate 같은 경우에는 서버를 실행시킬 수 있는 상태에서 테스트가 가능하지만
Controller를 직접 호출 하는 경우에는 서버의 구동 없이도 테스트가 가능합니다.

일반적으로 서비스가 볼륨을 키울수록 테스트 코드도 자연스레 커지게 되는데 (설정, 레거시 코드 등등),
그런 부분이 불편하다고 생각해 환경적인 요소를 배제하려고 의견을 드렸던 부분입니다.

yongdae reviewed (changes requested):
2단계 미션을 잘 진행해주고 계시네요. 👍 
아직 완성 되지 않았지만 설계를 잘 해주셔서 피드백을 드리기보다는 배운다는 생각으로 리뷰를 했네요.

질문 주신 부분은 포함해 몇 가지 의견을 남겨놓았습니다.
확인하시고 다시 리뷰 요청 부탁드리겠습니다 ~ 🔥

catsbi committed:
feat: 핵심 업무 규칙 추상화 및 로직 개선

- 매직 리터럴중 분리가 필요한 부분 분리
- StringBuilder 제거
- HttpHeaders 인스턴스 생성로직 변경
- JSON 변환을 위한 getter및 생성자 작성

catsbi committed:
refactor: 응답객체 개선(HttpResponse -> ResponseEntity)및 로직 개선

- View 객체 핵심업무로직 DefaultView로 집중 및 목적에 따른 구현체 분리
- index.html 정적 경로 절대경로로 변경
- 소스파일 reformat, import optimize

catsbi committed:
feat: 사용하지 않는 객체 제거 및 쿠키 테스트 케이스 작성

catsbi committed:
feat: 쿠키객체 및 DefaultView 객체 테스트 케이스 작성

catsbi committed:
refacotr: RequestBody 로직 개선 및 테스트 케이스 수정

lhs0610 committed:
refactor: 주석(comment) 작성

yongdae commented on src/test/java/webserver/domain/HttpHeadersTest.java:
```diff
@@ -0,0 +1,137 @@
+import static org.junit.jupiter.api.Assertions.*;
+import static webserver.domain.HttpHeaders.CONTENT_TYPE;
+
+class HttpHeadersTest {
```
요청/응답의 헤더를 하나의 객체로 잘 분리해주셨네요.
OutputStream을 포함해 세세하게 테스트를 만들어주셔서 좋았습니다. 💯

yongdae commented on src/test/java/webserver/HttpHeaderTest.java:
```diff
@@ -0,0 +1,57 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class HttpHeaderTest {
```
HttpHeadersTest와 중복되는 테스트가 있어보이는데 별개로 테스트를 하신 이유가 있으신가요?

catsbi replied:
그.........러게요?! 매일 퇴근하고 새벽진행을 하다보니, 정신이 없었던 것 같습니다 😥

yongdae commented on src/main/java/webserver/domain/Cookie.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Override
+    public String toString() {
```
로그인을 하면 Set-Cookie에 설정되는 값은 다음과 같습니다.

<img width="415" alt="스크린샷 2022-07-28 오후 10 05 06" src="https://user-images.githubusercontent.com/4317804/181512153-4923de47-ae69-4aa3-9cf8-bf3ae06ebe14.png">

쿠키 값이 `<cookie-name>=<cookie-value>; Path=<path-value>`
[이 규칙](https://developer.mozilla.org/ko/docs/Web/HTTP/Headers/Set-Cookie)을 따르지 않아 Cookie에 값이 올바르게 설정되지 않는 문제가 있네요.

catsbi replied:
확인 후 적용하겠습니다!

yongdae commented on src/main/java/webserver/ui/UserController.java:
```diff
@@ -0,0 +1,107 @@
+    private boolean isLogin(HttpRequest httpRequest) {
+        Cookie cookie = httpRequest.getCookie();
+        String loginedStr = cookie.getAttribute("logined");
+        return Boolean.getBoolean(loginedStr);
```
<img width="884" alt="스크린샷 2022-07-28 오후 10 16 36" src="https://user-images.githubusercontent.com/4317804/181514466-cc9622c1-85d0-40b6-ade8-10321a85fb2f.png">

확인해보니 Boolean.getBoolean()는 system property의 값을 읽어서
해당 값이 "true"인 경우에 true를 반환하는 것으로 판단됩니다.

여기서 Boolean.getBoolean()를 사용하신 이유가 있으신가요?

catsbi replied:
문자열을 논리값으로 변경하기 위해 사용했습니다. valueOf는 캐싱되어있는 TRUE, FALSE객체를 반환하는데 이는 wrapper타입이기에 unboxing에 걸리는 비용을 아끼려고 선택했는데 그냥 parseBoolean을 사용해도 될 것 같네요!

yongdae commented on src/main/java/webserver/domain/HttpStatus.java:
```diff
@@ -0,0 +1,31 @@
+    NOT_FOUND("404 Not Found"),
+    INTERNAL_SERVER_ERROR("500 Internal Server Error");
+
+    public static final String HTTP_1_1 = "HTTP/1.1 ";
```
여기에 Protocol과 Version을 사용해보면 어떨까요?

yongdae commented on src/main/java/webserver/domain/HttpRequest.java:
```diff
@@ -1,37 +1,86 @@
+     * @param br Http Request 정보를 담은 객체
+     * @return Http 요청 정보
+     */
+    public static HttpRequest newInstance(BufferedReader br) throws IOException {
```
👍

yongdae commented on src/main/java/webserver/domain/RequestMapping.java:
```diff
@@ -0,0 +1,19 @@
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(value = {ElementType.METHOD, ElementType.TYPE})
+public @interface RequestMapping {
```
@Retention, @Target를 사용해서 @RequestMapping 를 정의해주셨네요. 👍

yongdae commented on src/main/java/webserver/domain/ResponseEntity.java:
```diff
@@ -0,0 +1,148 @@
+        return status(HttpStatus.OK);
+    }
+
+    public static ResponseBuilder created(String locationValue) {
+        HttpHeaders httpHeaders = HttpHeaders.defaultResponseHeader();
+        httpHeaders.add(HttpHeaders.LOCATION, locationValue);
+
+        return status(HttpStatus.CREATED).headers(httpHeaders);
+    }
+
+    public static ResponseBuilder badRequest() {
+        return status(HttpStatus.BAD_REQUEST);
+    }
+
+    public static ResponseBuilder notFound() {
+        return status(HttpStatus.NOT_FOUND);
+    }
+
+    public static ResponseBuilder internalServerError() {
+        return status(HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+
+    public static ResponseBuilder found(String locationValue) {
+        HttpHeaders httpHeaders = HttpHeaders.defaultResponseHeader();
+        httpHeaders.add(HttpHeaders.LOCATION, locationValue);
+
+        return status(HttpStatus.FOUND).headers(httpHeaders);
+    }
+
+    public static ResponseBuilder status(HttpStatus status) {
+        return new ResponseBuilder<>(status);
+    }
```
ResponseEntity를 때에 따라 손쉽게 사용 할 수 있도록 정의해주셨네요. 💯

yongdae commented on src/main/java/webserver/domain/ResponseEntity.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+
+    public static class ResponseBuilder<T> {
```
👍

yongdae reviewed (approved):
이전에 의견드린 부분을 포함해 미션을 잘 수행해주셨네요. :)
조금 이슈가 있긴 하지만 큰 문제는 아니라고 판단해 바로 다음 단계로 넘어가도록 하겠습니다.
의견을 남겨 놓은 부분이 있으니 확인 부탁드릴께요. 수고하셨습니다!!
